### PR TITLE
Fix case where `get_partial_current_state_deltas` could return >100 rows

### PR DESCRIPTION
**Pull request by anoadragon453**
`get_partial_current_state_deltas` is supposed to return max 100 rows. We relied on this assumption in https://github.com/element-hq/synapse/pull/18926, forgoing batching of state deltas. It turns out that `get_partial_current_state_deltas` can potentially return millions of rows, and we saw this in production on element.io.

This PR fixes the case where every state delta in `current_state_delta_stream` has a count of 1, meaning `total` will be 100. Before this change, that would result in `clipped_stream_id = max_stream_id`, meaning we'd potentially pull out millions of rows.

~~I haven't written a regression test for this and I'm not sure how to easily seed `current_state_delta_stream` with the correct data. I suppose we could create a room and change the topic >100 times? Alternatively we could write directly to the table, but such a test would need to be updated if the schema changes.~~ I was able to write a unit test without writing SQL directly.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: tests/storage/test_state.py [hunk 1]
```diff
@@ -643,3 +643,101 @@ def test_batched_state_group_storing(self) -> None:
                     ),
                 )
                 self.assertEqual(context.state_group_before_event, groups[0][0])
+
+
+class CurrentStateDeltaStreamTestCase(HomeserverTestCase):
```
Comment (anoadragon453) [inline tests/storage/test_state.py:650]: This test works and fails prior to the fix commit. However, due to running `inject_state_event` >100 times, this test takes about ~3s to run on my mid-range laptop. I attempted to speed this up by batching events with the following code:

<details>

<summary>Click to show prior test code</summary>

```py
    def test_current_state_delta_stream_is_limited_to_100(self) -> None:
        # Fetch the current state group to pass into store_state_deltas_for_batched
        creation_event = self.inject_state_event(
            self.room, self.alice_user_id, EventTypes.Create, "", {}
        )
        state_to_event = self.get_success(
            self.storage.state.get_state_groups(
                self.room.to_string(), [creation_event.event_id]
            )
        )
        current_state_group = list(state_to_event.keys())[0]

        # Make >100 state changes for the same (type, state_key) to ensure many rows.
        # We’ll do 150 renames: this produces 150 rows in current_state_delta_stream.
        # TODO: Make this use the helper function.
        events_and_unpersisted_base_contexts: List[Tuple[EventBase, UnpersistedEventContextBase]] = []
        for i in range(150):
            builder = self.event_builder_factory.for_room_version(
                RoomVersions.V1,
                {
                    "type": EventTypes.Name,
                    "sender": self.alice_user_id.to_string(),
                    "state_key": "",
                    "room_id": self.room.to_string(),
                    "content": {"name": f"rename #{i}"},
                },
            )
            event, unpersisted_context = self.get_success(
                self.event_creation_handler.create_new_client_event(builder)
            )
            events_and_unpersisted_base_contexts.append((event, unpersisted_context))

        # Persist all those state deltas for the room in one batch (like production).
        events_and_persisted_context = self.get_success(
            UnpersistedEventContext.batch_persist_unpersisted_contexts(
                events_and_unpersisted_base_contexts,
                room_id=self.room.to_string(),
                last_known_state_group=current_state_group,
                datastore=self.state_datastore,
            )
        )
        self.get_success(
            self.storage.persistence.persist_events(events_and_persisted_context)
        )

        # Sanity check: count how many rows exist between prev=0 and max, it should be >100.
        max_stream_id = self.store.get_room_max_stream_ordering()

        # Call the function under test. With the >= 100 clipping fix, this must return <= 100 rows.
        clipped_stream_id, deltas = self.get_success(
            self.store.get_partial_current_state_deltas(prev_stream_id=0, max_stream_id=max_stream_id)
        )

        self.assertLessEqual(
            len(deltas), 100, f"Returned {len(deltas)} rows, expected at most 100"
        )

        # Optional: advancing from the clipped point should eventually drain the remainder.
        # Make sure we make progress and don’t get stuck.
        if deltas:
            next_prev = clipped_stream_id
            next_clipped, next_deltas = self.get_success(
                self.store.get_partial_current_state_deltas(
                    prev_stream_id=next_prev, max_stream_id=max_stream_id
                )
            )
            self.assertNotEqual(
                next_clipped, clipped_stream_id, "Did not advance clipped_stream_id"
            )
            # Still should respect the 100-row limit.
            self.assertLessEqual(len(next_deltas), 100)
```

</details>

Yet this resulted in the following entries for `current_state_delta_stream`:

```sql
sqlite> select * from current_state_delta_stream;
2|!abc1234:test|m.room.create||$00KVFIl:test||master
3|!abc1234:test|m.room.name||$041QljZO:test||master
103|!abc1234:test|m.room.name||$0132oOsiN:test|$041QljZO:test|master
```

which isn't what we're trying to test (>100 entries). So the test fails. I reverted to the working, slower code as a result. Any tips are appreciated, or we can run with this.

Comment (reivilibre) [reply tests/storage/test_state.py:650]: Can we add a `limit = 100` parameter and use a smaller limit in the test?

File: synapse/storage/databases/main/state_deltas.py [hunk 2]
```diff
@@ -133,7 +133,8 @@ def get_current_state_deltas_txn(
 
             for stream_id, count in txn:
```
Comment (reivilibre) [inline synapse/storage/databases/main/state_deltas.py]: wondering if instead of doing this with a for loop, we should just ask the DB for the result we want directly?

I think this would do it, as long as you still handle the case where there are no result rows (less than 100 left, so no need to clamp)

```sql
SELECT stream_id
FROM current_state_delta_stream
WHERE stream_id > ? AND stream_id <= ?
ORDER BY stream_id ASC
LIMIT 1 OFFSET 99
```

File: tests/storage/test_state.py (hunk 1 above)
Comment (anoadragon453) [reply tests/storage/test_state.py:650]: While this isn't ideal as we're just adding a parameter for the unit tests... I agree that this is a simple option that totally works. Thanks for the suggestion!

File: synapse/storage/databases/main/state_deltas.py (hunk 2 above)
Comment (anoadragon453) [reply synapse/storage/databases/main/state_deltas.py]: I actually just ended up using `LIMIT` in the second query in https://github.com/element-hq/synapse/pull/18960/commits/0e40bf4589c90294d45489bef9d72744bcbe6c7f.

Am I missing something - is it really this simple? All the unit tests appear to pass :shrug:

**Comment by anoadragon453** [comment]
Complement CI is failing due to [a flaky test](https://github.com/element-hq/synapse/issues/18537).

Edit: And sytest appears to just be stalling...

File: synapse/storage/databases/main/state_deltas.py [hunk 3]
```diff
@@ -114,46 +119,23 @@ async def get_partial_current_state_deltas(
         def get_current_state_deltas_txn(
             txn: LoggingTransaction,
         ) -> Tuple[int, List[StateDelta]]:
-            # First we calculate the max stream id that will give us less than
-            # N results.
-            # We limit the number of returned stream_id entries to ensure we
-            # don't select toooo many.
             sql = """
-                SELECT stream_id, count(*)
+                SELECT stream_id, room_id, type, state_key, event_id, prev_event_id
```
Comment (reivilibre) [inline synapse/storage/databases/main/state_deltas.py:176]: afraid this doesn't work AFAICT, because `stream_id` is not unique, but we always want to process the entire state with the same `stream_id` (never process some of the rows but not others)

**Review by reivilibre — CHANGES_REQUESTED** [review]
I mean.... it's weird if the tests don't catch this, probably a smell that we're missing a test

File: synapse/storage/databases/main/state_deltas.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/storage/databases/main/state_deltas.py:176]: So would this mean that if a group of state deltas with the same `stream_id` that was >100 items would cause this function to deadlock?

I suppose we should just return the whole group, even if it's larger than `limit`, in that case.

Edit: especially since the docstring states that if the list of returned state deltas is empty, then we are up to date. Potentially skipping over a lot of state deltas!

Comment (anoadragon453) [reply synapse/storage/databases/main/state_deltas.py:176]: I've implemented this in https://github.com/element-hq/synapse/pull/18960/commits/6d14263dea913949127398b3657c01790107c1cd, and added a test for the edge case in https://github.com/element-hq/synapse/pull/18960/commits/ae01153b88fad8cc313130f2ee7152023d0daef9.

Comment (reivilibre) [reply synapse/storage/databases/main/state_deltas.py:176]: I don't have memory of there being any deadlock here (in fact, given no concurrency primitives are involved — well except row locks? — it's hard to imagine this being on the table at all) but I think the problem was that a group with more than one row might not be covered within the `LIMIT 100` batch, in which case we would advance past its stream_id but not have processed all the rows within that group.
Even a group of 2 state deltas would be enough to trigger this, if it was at the 99th position in the limit, if you see what I mean.

But you also make a good point that we can indeed have groups larger than 100 deltas, I think because of state resolution (both desirable resolutions and state resets?), essentially.

File: synapse/storage/databases/main/state_deltas.py [hunk 4]
```diff
@@ -115,45 +129,71 @@ async def get_partial_current_state_deltas(
...
+                else:
+                    # Either we have already included some groups and adding
+                    # this one would exceed the limit, or this is the first
+                    # group and it alone exceeds the limit.
+                    hit_limit = True
+                    if included_rows == 0:
+                        # Return the entire oversized group so that callers make
+                        # progress (even though this may exceed `limit` rows).
+                        fetch_upto_stream_id = stream_id
                     break
```
Comment (reivilibre) [inline synapse/storage/databases/main/state_deltas.py]: up to you, but you may or may not like this alternative statement of this logic:

```suggestion
            # Always retrieve the first group, at the bare minimum
            fetch_upto_stream_id, included_rows = grouped_rows[0]
            # Determine which other groups we can retrieve at the same time,
            # without blowing the budget
            for stream_id, count in grouped_rows[1:]:
                if included_rows + count > limit:
                    break
                included_rows += count
                fetch_upto_stream_id = stream_id
```

(then also remove the if block beneath, since we don't need it any longer)
(probably should put a type annotation on `grouped_rows`)
(I think `hit_limit` is now unnecessary and you can just use `caught_up_with_stream = len(grouped_rows) < group_limit`?)

Sorry if the indentation is wrong, why does this use a variable-width textbox? :D

File: tests/storage/test_state.py [hunk 5]
```diff
@@ -643,3 +645,309 @@ def test_batched_state_group_storing(self) -> None:
...
+            len(next_deltas),
+            4,
+            f"Returned {len(next_deltas)} rows, expected all 4 m.room.name events: {next_deltas}",
+        )
+
+    def test_get_partial_current_state_deltas_does_not_deadlock(self) -> None:
+        """
+        Tests that `get_partial_current_state_deltas` does not repeatedly return
+        zero entries due to the passed `limit` parameter being less than the
+        size of the next group of state deltas from the given `prev_stream_id`.
```
Comment (reivilibre) [inline tests/storage/test_state.py]: ah, I see, `deadlock` is not (or at least I don't think it is) the same as returning zero entries.

To me deadlock is when multiple 'parties' hold locks in such a way that all of them become dead whilst waiting on others to release a lock. Maybe easier illustrated with pseudocode:
```
process A {
    lock(X);
    // <---
    lock(Y);
}
```

```
process B {
	lock(Y);
    // <---
	lock(X);
}
```

If both processes make it to the commented arrow, neither can obtain the second lock, so both wait indefinitely and 'deadlock'.

(There is also 'livelock' where multiple 'parties' are spinning away, being active/a**live** whilst somehow conflicting with each other in a way that means they all fail to advance. I don't remember seeing this in real life yet, so I'm going to say it's more niche.)

File: tests/storage/test_state.py [hunk 6]
```diff
@@ -643,3 +645,309 @@ def test_batched_state_group_storing(self) -> None:
...
+        # 1 2 3 4 5
+        #
+        # C = m.room.create
+        # T = m.room.topic
+        #
+        # With the limit, we should return only the create event, as returning 4
+        # state deltas would result in splitting a group:
+        #
+        # C T T T T
+        # 1 2 3 4 X
```
Comment (reivilibre) [inline tests/storage/test_state.py:939]: Copypaste that is stale for this example, I think :)

**Review by reivilibre — APPROVED** [review]
Looks correct, some small bits here and there but optional really

File: tests/storage/test_state.py (hunk 6 above)
Comment (anoadragon453) [reply tests/storage/test_state.py:939]: Actually, this is relevant! But the description is a little off (they are `m.room.name` events, not topic events). Plus the commented out assertion below is definitely not supposed to be there. Looks like these tests were indeed still WIP.

Fixed up in 0712b31be56a5da30a511588208fdf7efe122d4f

File: tests/storage/test_state.py (hunk 5 above)
Comment (anoadragon453) [reply tests/storage/test_state.py]: Good point on the terminology. Perhaps "an infinite loop" is most appropriate (https://github.com/element-hq/synapse/pull/18960/commits/aadd2d8c94f7bb65d1a28ecc36317f864ccd803d)?

File: synapse/storage/databases/main/state_deltas.py (hunk 4 above)
Comment (anoadragon453) [reply synapse/storage/databases/main/state_deltas.py]: Ooo, yes, that's much cleaner. Thank you!

**Review by anoadragon453 — COMMENTED** [review]
Thanks for the review (and code suggestions) @reivilibre. Super helpful!
